src/main/java/org/gradle/profiler/studio/StudioGradleClient.java

```diff
@@ -71,23 +82,48 @@ public BuildActionResult sync(List<String> gradleArgs, List<String> jvmArgs) {
-            System.out.println("* Full sync has completed in: " + syncRequestCompleted.getDurationMillis() + "ms and it " + syncRequestCompleted.getResult().name().toLowerCase());
-            return BuildActionResult.withIdeTimings(
-                Duration.ofMillis(syncRequestCompleted.getDurationMillis()),
-                Duration.ofMillis(agentCompleted.getDurationMillis()),
-                Duration.ofMillis(syncRequestCompleted.getDurationMillis() - agentCompleted.getDurationMillis())
-            );
+            Pair<StudioSyncRequestResult, BuildActionResult> pair = waitForSyncToFinish(connections, gradleArgs, jvmArgs);
+            StudioSyncRequestResult syncRequestResult = pair.getLeft();
+            BuildActionResult durationResult = pair.getRight();
+            System.out.printf("* Full Gradle execution time: %dms%n", durationResult.getGradleToolingAgentExecutionTime().get().toMillis());
+            System.out.printf("* Full IDE execution time: %dms%n", durationResult.getIdeExecutionTime().get().toMillis());
+            System.out.printf("* Full sync has completed in: %dms and it %s%n", durationResult.getExecutionTime().toMillis(), syncRequestResult);
+            return durationResult;
         });
     }
 
+    private Pair<StudioSyncRequestResult, BuildActionResult> waitForSyncToFinish(StudioConnections connections, List<String> gradleArgs, List<String> jvmArgs) {
+        System.out.println("* Sync has started, waiting for it to complete...");
+        AtomicBoolean isSyncRequestCompleted = new AtomicBoolean();
+        CompletableFuture<Long> gradleInvocations = CompletableFuture.supplyAsync(() -> collectGradleInvocations(connections, isSyncRequestCompleted, gradleArgs, jvmArgs), executor);
+        StudioSyncRequestCompleted syncRequestCompleted = connections.getPluginConnection().receiveSyncRequestCompleted(SYNC_REQUEST_COMPLETED_TIMEOUT);
+        isSyncRequestCompleted.set(true);
+        long gradleInvocationsDuration = gradleInvocations.join();
+        BuildActionResult result = BuildActionResult.withIdeTimings(
+            Duration.ofMillis(syncRequestCompleted.getDurationMillis()),
+            Duration.ofMillis(gradleInvocationsDuration),
+            Duration.ofMillis(syncRequestCompleted.getDurationMillis() - gradleInvocationsDuration)
+        );
+        return Pair.of(syncRequestCompleted.getResult(), result);
+    }
+
+    private long collectGradleInvocations(StudioConnections connections, AtomicBoolean isSyncRequestCompleted, List<String> gradleArgs, List<String> jvmArgs) {
+        long durations = 0;
+        int invocation = 1;
+        ServerConnection agentConnection = connections.getAgentConnection();
+        while (!isSyncRequestCompleted.get()) {
+            Optional<GradleInvocationStarted> invocationStarted = agentConnection.maybeReceiveGradleInvocationStarted(GRADLE_INVOCATION_STARTED_TIMEOUT);
```

Comment:
Would it make sense here to abort the waiting when the sync finishes?

Reply:
For that reason we have GRADLE_INVOCATION_STARTED_TIMEOUT set to low value 2 seconds atm, so we regularly check if `isSyncRequestCompleted` is done. I will lower that value some more, e.g. 100ms.

Problem I was facing was, that interrupting reading from InputStream was causing that further reads being blocked or broken.